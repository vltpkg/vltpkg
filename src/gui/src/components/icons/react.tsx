export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
