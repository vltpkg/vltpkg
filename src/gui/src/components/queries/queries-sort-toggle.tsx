import { type Option, Toggle } from '@/components/ui/toggle.jsx'
import { ArrowDownAz, ArrowDownZa } from 'lucide-react'

export type SortingOption = 'ascending' | 'descending'

interface QueriesSortToggleProps<T> {
  filteredItems: T[]
  setFilteredItems: (i: T[]) => void
  sortKey: keyof T
}

interface SortOption extends Option {
  key: SortingOption
}

export const sortAlphabeticallyAscending = <T,>(
  items: T[],
  sortKey: keyof T,
  setItems: (i: T[]) => void,
) => {
  const sorted = [...items].sort((a, b) =>
    String(a[sortKey]).localeCompare(String(b[sortKey])),
  )
  setItems(sorted)
}

export const sortAlphabeticallyDescending = <T,>(
  items: T[],
  sortKey: keyof T,
  setItems: (i: T[]) => void,
) => {
  const sorted = [...items].sort((a, b) =>
    String(b[sortKey]).localeCompare(String(a[sortKey])),
  )
  setItems(sorted)
}

const QueriesSortToggle = <T,>({
  filteredItems,
  setFilteredItems,
  sortKey,
}: QueriesSortToggleProps<T>) => {
  const options: [SortOption, SortOption] = [
    {
      icon: props => <ArrowDownAz {...props} />,
      toolTipContent: 'Ascending',
      key: 'ascending',
      callBack: () =>
        sortAlphabeticallyAscending<T>(
          filteredItems,
          sortKey,
          setFilteredItems,
        ),
    },
    {
      icon: props => <ArrowDownZa {...props} />,
      toolTipContent: 'Descending',
      key: 'descending',
      callBack: () =>
        sortAlphabeticallyDescending<T>(
          filteredItems,
          sortKey,
          setFilteredItems,
        ),
    },
  ]

  return <Toggle options={options} />
}

export { QueriesSortToggle }
