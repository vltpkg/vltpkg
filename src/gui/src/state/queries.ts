import { create } from 'zustand'
import type { SavedQuery } from '@/state/types.ts'

export type QueriesState = {
  isCreating: boolean
  deleteDialogOpen: boolean
  filteredQueries: SavedQuery[]
  selectedQueries: SavedQuery[]
}

export type QueriesAction = {
  setIsCreating: (isCreating: QueriesState['isCreating']) => void
  setDeleteDialogOpen: (
    deleteDialogOpen: QueriesState['deleteDialogOpen'],
  ) => void
  setFilteredQueries: (
    filteredQueries: QueriesState['filteredQueries'],
  ) => void
  setSelectedQueries: (
    selectedQueries: QueriesState['selectedQueries'],
  ) => void
}

const initialState: QueriesState = {
  isCreating: false,
  deleteDialogOpen: false,
  filteredQueries: [],
  selectedQueries: [],
}

export const useQueriesStore = create<QueriesState & QueriesAction>(
  set => {
    const store = {
      ...initialState,
      setIsCreating: (isCreating: QueriesState['isCreating']) =>
        set(() => ({ isCreating })),
      setFilteredQueries: (
        filteredQueries: QueriesState['filteredQueries'],
      ) => set(() => ({ filteredQueries })),
      setDeleteDialogOpen: (
        deleteDialogOpen: QueriesState['deleteDialogOpen'],
      ) => set(() => ({ deleteDialogOpen })),
      setSelectedQueries: (
        selectedQueries: QueriesState['selectedQueries'],
      ) => set(() => ({ selectedQueries })),
    }

    return store
  },
)
