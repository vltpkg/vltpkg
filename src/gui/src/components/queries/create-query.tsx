import { useEffect, useState } from 'react'
import { Button } from '@/components/ui/button.tsx'
import { Input } from '@/components/ui/input.tsx'
import { useGraphStore } from '@/state/index.ts'
import { v4 as uuidv4 } from 'uuid'
import type { QueryLabel, DashboardData } from '@/state/types.ts'
import { LabelSelect } from '@/components/labels/label-select.tsx'
import { LabelBadge } from '@/components/labels/label-badge.tsx'
import { useToast } from '@/components/hooks/use-toast.ts'
import { Label } from '@/components/ui/label.tsx'
import {
  Popover,
  PopoverTrigger,
  PopoverContent,
} from '@/components/ui/popover.tsx'
import {
  Tooltip,
  TooltipProvider,
  TooltipTrigger,
  TooltipContent,
} from '@/components/ui/tooltip.tsx'
import { ChevronsUpDown, CircleHelp } from 'lucide-react'
import { cn } from '@/lib/utils.ts'
import { DirectorySelect } from '@/components/directory-select.tsx'

interface CreateQueryProps {
  onClose: () => void
  className?: string
  dashboard?: DashboardData
}

export const CreateQuery = ({
  className,
  onClose,
  dashboard,
}: CreateQueryProps) => {
  const [queryName, setQueryName] = useState<string>('')
  const [query, setQuery] = useState<string>('')
  const [directory, setDirectory] = useState<string>('')
  const [labels, setLabels] = useState<QueryLabel[]>([])

  const [isCreationValid, setIsCreationValid] =
    useState<boolean>(false)
  const [_labelPopoverOpen, setLabelPopoverOpen] =
    useState<boolean>(false)

  const { toast } = useToast()
  const saveQuery = useGraphStore(state => state.saveQuery)

  useEffect(() => {
    if (queryName.trim() !== '' && query.trim() !== '') {
      setIsCreationValid(true)
    } else {
      setIsCreationValid(false)
    }
  }, [queryName, query])

  const handleCancel = () => {
    onClose()
  }

  const handleCreate = () => {
    saveQuery({
      id: uuidv4(),
      name: queryName,
      context: directory,
      query,
      dateCreated: new Date().toISOString(),
      dateModified: new Date().toISOString(),
      labels,
    })
    toast({
      title: `${queryName} saved successfully`,
    })
    onClose()
  }

  return (
    <div
      className={cn(
        'rounded-sm border border-[1px] border-muted-foreground/25 bg-card transition-all',
        className,
      )}>
      <div className="flex flex-col gap-3 px-3 py-3">
        {/* query preview */}
        <div className="flex w-full items-center">
          <p className="text-base font-medium">
            {queryName.trim() === '' ? 'New Query' : queryName}
          </p>

          <div className="ml-auto flex items-center gap-2">
            {labels.map((label, idx) => (
              <LabelBadge key={idx} {...label} />
            ))}
          </div>
        </div>

        <div className="mt-4 flex w-full gap-3">
          <div className="flex flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Name
            </Label>
            <Input
              type="text"
              value={queryName}
              onChange={e => setQueryName(e.target.value)}
              placeholder="Name"
            />
          </div>

          <div className="flex grow flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Query
            </Label>
            <Input
              type="text"
              value={query}
              onChange={e => setQuery(e.target.value)}
              placeholder="Query"
            />
          </div>

          <div className="flex w-[300px] flex-col gap-2">
            <TooltipProvider>
              <Tooltip>
                <TooltipTrigger className="inline-flex cursor-default items-center">
                  <Label className="border-none text-sm font-medium">
                    Directory (optional)
                  </Label>
                  <CircleHelp
                    className="text-muted-foreground"
                    size={18}
                  />
                </TooltipTrigger>
                <TooltipContent>
                  Set directory to 'Global' to reuse across all
                  projects.
                </TooltipContent>
              </Tooltip>
            </TooltipProvider>
            <DirectorySelect
              directory={directory}
              setDirectory={setDirectory}
              dashboard={dashboard}
            />
          </div>

          <div className="flex w-[175px] flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Labels
            </Label>
            <Popover>
              <PopoverTrigger asChild>
                <Button
                  variant="outline"
                  role="combobox"
                  className="w-full justify-between">
                  Select labels
                  <ChevronsUpDown className="opacity-50" />
                </Button>
              </PopoverTrigger>
              <PopoverContent className="p-0">
                <LabelSelect
                  setIsOpen={setLabelPopoverOpen}
                  selectedItems={labels}
                  setItems={setLabels}
                />
              </PopoverContent>
            </Popover>
          </div>
        </div>

        {/* footer */}
        <div className="mb-0.5 flex items-end justify-end gap-3">
          <Button onClick={handleCancel} variant="outline">
            Cancel
          </Button>
          <Button disabled={!isCreationValid} onClick={handleCreate}>
            Create Query
          </Button>
        </div>
      </div>
    </div>
  )
}
