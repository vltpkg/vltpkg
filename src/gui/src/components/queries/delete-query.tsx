import { useState } from 'react'
import {
  Dialog,
  DialogTitle,
  DialogTrigger,
  DialogDescription,
  DialogHeader,
  DialogFooter,
  DialogContent,
} from '@/components/ui/dialog.jsx'
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
} from '@/components/ui/tooltip.jsx'
import { Button } from '@/components/ui/button.jsx'
import { type SavedQuery } from '@/state/types.js'
import { Trash } from 'lucide-react'
import { useToast } from '@/components/hooks/use-toast.js'
import { useGraphStore } from '@/state/index.js'

interface DeleteQueryProps {
  deleteDialogOpen: boolean
  setDeleteDialogOpen: (o: boolean) => void
  selectedQueries: SavedQuery[]
  type: 'button' | 'icon'
  onClose?: () => void
  text?: string
}

const DeleteQuery = ({
  deleteDialogOpen,
  setDeleteDialogOpen,
  selectedQueries,
  type,
  onClose,
  text = 'Delete',
}: DeleteQueryProps) => {
  const { toast } = useToast()
  const [isTooltipOpen, setIsTooltipOpen] = useState<boolean>(false)
  const deleteQueries = useGraphStore(
    state => state.deleteSavedQueries,
  )

  const confirmDelete = () => {
    setDeleteDialogOpen(false)
    setIsTooltipOpen(false)
    deleteQueries(selectedQueries)
    toast({
      title: 'Deleted successfully',
    })
    if (onClose) onClose()
  }

  return (
    <Dialog
      open={deleteDialogOpen}
      onOpenChange={setDeleteDialogOpen}>
      <Tooltip open={isTooltipOpen} onOpenChange={setIsTooltipOpen}>
        <TooltipProvider>
          <DialogTrigger asChild>
            {type === 'icon' ?
              <Button
                onMouseEnter={() => setIsTooltipOpen(true)}
                onMouseLeave={() => setIsTooltipOpen(false)}
                disabled={selectedQueries.length <= 0}
                className="rounded-sm border border-[1px] border-muted-foreground/25"
                size="icon"
                variant="ghost">
                <Trash />
              </Button>
            : <Button variant="destructive">{text}</Button>}
          </DialogTrigger>
          <TooltipContent>
            Delete {selectedQueries.length <= 1 ? 'Query' : 'Queries'}
          </TooltipContent>
        </TooltipProvider>
      </Tooltip>
      <DialogContent>
        <DialogHeader>
          <DialogTitle className="font-medium">
            Delete {selectedQueries.length <= 1 ? 'Query' : 'Queries'}
          </DialogTitle>
          <DialogDescription />
        </DialogHeader>
        <div>
          You're about to delete:
          <ul className="my-3 list-none">
            {selectedQueries.map((query, idx) => (
              <li key={idx} className="font-medium">
                {query.name}
              </li>
            ))}
          </ul>
          This action cannot be undone.
        </div>
        <DialogFooter>
          <Button
            variant="outline"
            onClick={() => setDeleteDialogOpen(false)}>
            No, Keep It.
          </Button>
          <Button variant="destructive" onClick={confirmDelete}>
            Yes, Delete!
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  )
}

export { DeleteQuery }
