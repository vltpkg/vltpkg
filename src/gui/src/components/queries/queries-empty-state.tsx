import { useEffect } from 'react'
import { Button } from '@/components/ui/button.jsx'
import { ArrowRight, Command, Star } from 'lucide-react'
import { motion, useAnimate } from 'framer-motion'

const QueriesEmptyState = () => {
  const [starScope, starAnimate] = useAnimate<HTMLDivElement>()
  const [cursorScope, cursorAnimate] = useAnimate<HTMLDivElement>()

  useEffect(() => {
    const {
      x: cursorX,
      y: cursorY,
      width: cursorWidth,
      height: cursorHeight,
    } = cursorScope.current.getBoundingClientRect()

    const cursorInitialPosition = {
      centerX: cursorX + cursorWidth / 2,
      centerY: cursorY + cursorHeight / 2,
    }

    const {
      x: starX,
      y: starY,
      width: starWidth,
      height: starHeight,
    } = starScope.current.getBoundingClientRect()

    const initialStarPosition = {
      centerX: starX + starWidth,
      centerY: starY + starHeight,
    }

    cursorAnimate(
      cursorScope.current,
      {
        x: [
          0,
          -(
            cursorInitialPosition.centerX -
            initialStarPosition.centerX
          ),
          0,
        ],
        y: [
          55,
          -(
            cursorInitialPosition.centerY -
            initialStarPosition.centerY
          ),
          0,
        ],
      },
      {
        duration: 2,
      },
    )
    starAnimate(
      starScope.current,
      {
        scale: [1, 1.2, 1],
      },
      {
        delay: 0.75,
        duration: 0.5,
      },
    )
    starAnimate(
      '#icon',
      {
        rotate: [0, -71],
      },
      {
        delay: 0.75,
        duration: 0.5,
      },
    )
    starAnimate(
      '#icon',
      {
        fill: ['', 'currentColor'],
      },
      {
        delay: 0.75,
        duration: 0.5,
      },
    )

    const timeout = setTimeout(() => {
      cursorAnimate(
        cursorScope.current,
        {
          x: [0, Math.random() * 10, Math.random() * 10],
          y: [0, Math.random() * 10, Math.random() * 10],
        },
        {
          ease: 'easeInOut',
          duration: 2,
          repeat: Infinity,
          repeatType: 'mirror',
        },
      )
    }, 2500)

    return () => {
      clearTimeout(timeout)
    }
  }, [starScope, cursorScope])

  return (
    <div className="flex h-full w-full items-center justify-center bg-white dark:bg-black">
      <div className="-mt-20 mt-10 flex flex-col items-center justify-center gap-3">
        <div className="relative mb-20 flex items-center justify-center">
          <motion.div
            initial={{ opacity: 1 }}
            animate={{ opacity: [0, 1, 0] }}
            transition={{ duration: 3, repeat: Infinity, delay: 0.5 }}
            className="absolute z-[1] h-[450px] w-[450px] rounded-full border-[1px] border-neutral-500/10"
          />
          <motion.div
            initial={{ opacity: 1 }}
            animate={{ opacity: [0, 1, 0] }}
            transition={{ duration: 3, repeat: Infinity, delay: 1 }}
            className="absolute z-[1] h-[300px] w-[300px] rounded-full border-[1px] border-neutral-500/15"
          />
          <motion.div
            initial={{ opacity: 1 }}
            animate={{ opacity: [0, 1, 0] }}
            transition={{ duration: 3, repeat: Infinity, delay: 1.5 }}
            className="absolute z-[1] h-[150px] w-[150px] rounded-full border-[1px] border-neutral-500/20"
          />

          <div className="absolute z-[2] flex h-[50px] w-[400px] items-center justify-center rounded-[8px] bg-gradient-to-r from-white to-neutral-400 dark:from-black dark:to-neutral-800">
            <div className="relative flex h-[48.25px] w-[398.25px] items-center rounded-[6.25px] bg-white shadow-2xl dark:bg-black">
              <p className="absolute -ml-32 whitespace-nowrap bg-gradient-to-r from-white to-neutral-500 bg-clip-text text-transparent dark:from-black dark:to-neutral-500">{`[name="my-project"] > :is(:project > *[name="tap"])`}</p>
              <div className="ml-auto mr-3 flex gap-2">
                <div
                  ref={starScope}
                  className="flex h-[24px] w-[24px] items-center justify-center rounded-sm border-[1px] border-neutral-800 bg-neutral-700 dark:border-neutral-900 dark:bg-neutral-300">
                  <Star
                    id="icon"
                    size={16}
                    className="text-neutral-100 dark:text-neutral-900"
                  />
                </div>
                <div className="flex items-center justify-center gap-1 before:mr-1 before:h-[14px] before:w-[0.9px] before:rounded-full before:bg-neutral-400 before:content-[''] dark:before:bg-neutral-800">
                  <div className="flex h-[24px] w-[24px] items-center justify-center rounded-sm border-[1px] border-neutral-400 bg-neutral-100 dark:border-neutral-800 dark:bg-neutral-900">
                    <Command size={13} className="text-neutral-500" />
                  </div>
                  <div className="flex h-[24px] w-[24px] items-center justify-center rounded-sm border-[1px] border-neutral-400 bg-neutral-100 dark:border-neutral-800 dark:bg-neutral-900">
                    <span className="font-mono text-sm text-neutral-500">
                      k
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              ref={cursorScope}
              className="absolute -right-10 top-10">
              <div className="relative">
                <svg
                  stroke="currentColor"
                  fill="currentColor"
                  strokeWidth="1"
                  viewBox="0 0 16 16"
                  className="h-6 w-6 -translate-x-[12px] -translate-y-[10px] -rotate-[70deg] transform stroke-sky-500 text-sky-400 dark:stroke-sky-500 dark:text-sky-700"
                  height="1em"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg">
                  <path d="M14.082 2.182a.5.5 0 0 1 .103.557L8.528 15.467a.5.5 0 0 1-.917-.007L5.57 10.694.803 8.652a.5.5 0 0 1-.006-.916l12.728-5.657a.5.5 0 0 1 .556.103z"></path>
                </svg>
              </div>
            </div>
          </div>
        </div>
        <h3 className="text-xl font-medium">No saved Queries yet</h3>
        <p className="text-center text-sm text-neutral-500">
          Looks like you haven't saved any Queries yet,
          <br />
          explore your Projects to add one!
        </p>
        <div className="flex gap-3"></div>
        <Button asChild className="z-[2]">
          <a href="/">
            <span>Explore Projects</span>
            <ArrowRight />
          </a>
        </Button>
      </div>
    </div>
  )
}

export { QueriesEmptyState }
