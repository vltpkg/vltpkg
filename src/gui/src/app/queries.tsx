import { useEffect, useState } from 'react'
import { useGraphStore } from '@/state/index.js'
import { type SavedQuery } from '@/state/types.js'
import { SavedQueryItem } from '@/components/queries/saved-item.jsx'
import { FilterSearch } from '@/components/ui/filter-search.jsx'
import { DeleteQuery } from '@/components/queries/delete-query.jsx'
import {
  sortAlphabeticallyAscending,
  SortToggle,
} from '@/components/sort-toggle.jsx'
import { Button } from '@/components/ui/button.jsx'
import { Tag } from 'lucide-react'
import { Badge } from '@/components/ui/badge.jsx'
import { Checkbox } from '@/components/ui/checkbox.jsx'
import { QueriesEmptyState } from '@/components/queries/queries-empty-state.jsx'

const Queries = () => {
  const savedQueries = useGraphStore(state => state.savedQueries)
  const [filteredQueries, setFilteredQueries] = useState<
    SavedQuery[]
  >(savedQueries ?? [])
  const [deleteDialogOpen, setDeleteDialogOpen] =
    useState<boolean>(false)
  const [selectedQueries, setSelectedQueries] = useState<
    SavedQuery[]
  >([])
  const savedLabels = useGraphStore(state => state.savedQueryLabels)

  const handleSelectQuery = (selectedQuery: SavedQuery) => {
    setSelectedQueries(prev => {
      const isSelected = prev.some(
        query => query.id === selectedQuery.id,
      )
      return isSelected ?
          prev.filter(query => query.id !== selectedQuery.id)
        : [...prev, selectedQuery]
    })
  }

  const handleSelectAll = () => {
    if (filteredQueries.length === selectedQueries.length) {
      setSelectedQueries([])
    } else {
      setSelectedQueries(filteredQueries)
    }
  }

  useEffect(() => {
    if (savedQueries) {
      sortAlphabeticallyAscending(
        filteredQueries,
        'name',
        setFilteredQueries,
      )
    }
  }, [savedQueries])

  return (
    <section className="flex min-h-[80svh] w-full grow flex-col bg-white dark:bg-black">
      <div className="flex flex-col px-8 py-4">
        <div className="flex justify-between">
          <div className="flex gap-2">
            <FilterSearch
              placeholder="Filter Queries"
              items={savedQueries}
              setFilteredItems={setFilteredQueries}
            />
            <SortToggle
              filteredItems={filteredQueries}
              setFilteredItems={setFilteredQueries}
              sortKey="name"
            />
            <div className="flex items-center">
              <div className="flex items-center">
                <DeleteQuery
                  deleteDialogOpen={deleteDialogOpen}
                  setDeleteDialogOpen={setDeleteDialogOpen}
                  selectedQueries={selectedQueries}
                  type="icon"
                />
              </div>
            </div>
          </div>
          <Button asChild variant="outline">
            <a href="/labels">
              <Tag />
              <span>Labels</span>
              <Badge variant="secondary">{savedLabels?.length}</Badge>
            </a>
          </Button>
        </div>

        <div className="mt-6">
          {!!savedQueries?.length && (
            <div className="grid grid-cols-12 gap-4">
              <div className="col-span-2 flex items-center gap-3 px-3">
                <Checkbox
                  onCheckedChange={handleSelectAll}
                  className="border-muted-foreground/50 hover:border-muted-foreground/75"
                />
                <p className="text-sm font-medium text-neutral-500">
                  Name
                </p>
              </div>
              <div className="col-span-4 items-center px-1">
                <p className="text-sm font-medium text-neutral-500">
                  Query
                </p>
              </div>
              <div className="col-span-2 items-center">
                <p className="text-sm font-medium text-neutral-500">
                  Directory
                </p>
              </div>
              <div className="col-span-4 items-center">
                <p className="text-sm font-medium text-neutral-500">
                  Labels
                </p>
              </div>
            </div>
          )}

          <div className="mt-3 flex flex-col gap-3">
            {savedQueries &&
              filteredQueries.map((query, idx) => (
                <SavedQueryItem
                  checked={selectedQueries.some(
                    selected => selected.id === query.id,
                  )}
                  handleSelect={handleSelectQuery}
                  key={idx}
                  item={query}
                />
              ))}
          </div>
        </div>
      </div>

      {!savedQueries?.length && <QueriesEmptyState />}
    </section>
  )
}

export { Queries }
