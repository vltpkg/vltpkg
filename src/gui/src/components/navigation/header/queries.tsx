import { NavLink } from 'react-router'
import { FilterSearch } from '@/components/ui/filter-search.tsx'
import { DeleteQuery } from '@/components/queries/delete-query.tsx'
import { Button } from '@/components/ui/button.tsx'
import { Plus, Tag } from 'lucide-react'
import { Badge } from '@/components/ui/badge.tsx'
import { SortToggle } from '@/components/sort-toggle.tsx'
import { useGraphStore } from '@/state/index.ts'
import { useQueriesStore } from '@/state/queries.ts'

export const QueriesHeader = () => {
  const savedLabels = useGraphStore(state => state.savedQueryLabels)
  const savedQueries = useGraphStore(state => state.savedQueries)

  const filteredQueries = useQueriesStore(
    state => state.filteredQueries,
  )
  const setFilteredQueries = useQueriesStore(
    state => state.setFilteredQueries,
  )
  const isCreating = useQueriesStore(state => state.isCreating)
  const setIsCreating = useQueriesStore(state => state.setIsCreating)
  const deleteDialogOpen = useQueriesStore(
    state => state.deleteDialogOpen,
  )
  const setDeleteDialogOpen = useQueriesStore(
    state => state.setDeleteDialogOpen,
  )
  const selectedQueries = useQueriesStore(
    state => state.selectedQueries,
  )

  return (
    <div className="flex w-full items-center justify-between gap-5">
      <FilterSearch
        placeholder="Filter Queries"
        items={savedQueries}
        className="grow"
        setFilteredItems={setFilteredQueries}
      />
      <SortToggle
        filteredItems={filteredQueries}
        setFilteredItems={setFilteredQueries}
        sortKey="name"
        classNames={{
          toggleClassName: 'rounded-xl',
          sliderClassName: 'rounded-[9px]',
          optionClassName: 'rounded-lg',
        }}
      />
      <div className="flex items-center">
        <div className="flex items-center">
          <DeleteQuery
            deleteDialogOpen={deleteDialogOpen}
            setDeleteDialogOpen={setDeleteDialogOpen}
            selectedQueries={selectedQueries}
            className="rounded-xl"
            type="icon"
          />
        </div>
      </div>
      {savedQueries?.length !== 0 && (
        <Button
          disabled={isCreating}
          onClick={() => setIsCreating(true)}
          className="rounded-xl">
          <span className="hidden md:flex">New Query</span>
          <Plus />
        </Button>
      )}
      <Button asChild variant="outline" className="rounded-xl">
        <NavLink to="/labels">
          <Tag />
          <span>Labels</span>
          <Badge variant="secondary">{savedLabels?.length}</Badge>
        </NavLink>
      </Button>
    </div>
  )
}
