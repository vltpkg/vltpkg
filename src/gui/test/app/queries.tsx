import { vi, expect, afterEach, describe, it } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { Queries } from '@/app/queries.jsx'

vi.mock('@/components/queries/saved-item.jsx', () => ({
  SavedQueryItem: 'gui-saved-query-item',
}))

vi.mock('@/components/ui/filter-search.jsx', () => ({
  FilterSearch: 'gui-filter-search',
}))

vi.mock('@/components/queries/delete-query.jsx', () => ({
  DeleteQuery: 'gui-delete-query',
}))

vi.mock('@/components/sort-toggle.jsx', () => ({
  SortToggle: 'gui-sort-toggle',
  sortAlphabeticallyAscending: vi.fn(),
}))

vi.mock('@/components/ui/button.jsx', () => ({
  Button: 'gui-button',
}))

vi.mock('lucide-react', () => ({
  Tag: 'gui-tag-icon',
}))

vi.mock('@/components/ui/badge.jsx', () => ({
  Badge: 'gui-badge',
}))

vi.mock('@/components/ui/checkbox.jsx', () => ({
  Checkbox: 'gui-checkbox',
}))

vi.mock('@/components/queries/queries-empty-state.jsx', () => ({
  QueriesEmptyState: 'gui-queries-empty-state',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

describe('labels view', () => {
  it('labels render default', () => {
    const Container = () => {
      return <Queries />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })
})
