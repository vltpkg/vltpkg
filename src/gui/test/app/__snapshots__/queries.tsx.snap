// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`labels view > labels render default 1`] = `

<section class="flex min-h-[80svh] w-full grow flex-col bg-white dark:bg-black">
  <div class="flex flex-col px-8 py-4">
    <div class="flex justify-between">
      <div class="flex gap-2">
        <gui-filter-search
          placeholder="Filter Queries"
          items
        >
        </gui-filter-search>
        <gui-sort-toggle
          filtereditems
          sortkey="name"
        >
        </gui-sort-toggle>
        <div class="flex items-center">
          <div class="flex items-center">
            <gui-delete-query
              deletedialogopen="false"
              selectedqueries
              type="icon"
            >
            </gui-delete-query>
          </div>
        </div>
      </div>
      <gui-button
        aschild="true"
        variant="outline"
      >
        <a href="/labels">
          <gui-tag-icon>
          </gui-tag-icon>
          <span>
            Labels
          </span>
          <gui-badge variant="secondary">
            3
          </gui-badge>
        </a>
      </gui-button>
    </div>
    <div class="mt-6">
      <div class="mt-3 flex flex-col gap-3">
      </div>
    </div>
  </div>
  <gui-queries-empty-state>
  </gui-queries-empty-state>
</section>

`;
